xrdp sessions fail with Quest/OneIdentity Safeguard for Privileged Sessions

### xrdp version

0.10.2

### Detailed xrdp version, build options

~~~
xrdp 0.10.2
  A Remote Desktop Protocol Server.
  Copyright (C) 2004-2024 Jay Sorg, Neutrino Labs, and all contributors.
  See https://github.com/neutrinolabs/xrdp for more information.

  Configure options:
      --build=x86_64-redhat-linux-gnu
      --host=x86_64-redhat-linux-gnu
      --program-prefix=
      --disable-dependency-tracking
      --prefix=/usr
      --exec-prefix=/usr
      --bindir=/usr/bin
      --sbindir=/usr/sbin
      --sysconfdir=/etc
      --datadir=/usr/share
      --includedir=/usr/include
      --libdir=/usr/lib64
      --libexecdir=/usr/libexec
      --localstatedir=/var
      --sharedstatedir=/var/lib
      --mandir=/usr/share/man
      --infodir=/usr/share/info
      --enable-fuse
      --enable-pixman
      --enable-painter
      --enable-vsock
      --enable-ipv6
      --enable-utmp
      --with-socketdir=/run/xrdp
      --with-imlib2
      build_alias=x86_64-redhat-linux-gnu
      host_alias=x86_64-redhat-linux-gnu
      CFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  
      LDFLAGS=-Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld
      CXXFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection
      PKG_CONFIG_PATH=:/usr/lib64/pkgconfig:/usr/share/pkgconfig

  Compiled with OpenSSL 1.1.1k  FIPS 25 Mar 2021
~~~



### Operating system & version

Oracle Linux 8.9

### Installation method

dnf / apt / zypper / pkg / etc

### Which backend do you use?

Xvnc (tigervnc-server-minimal-1.13.1-15.el8_10)

### What desktop environment do you use?

any

### Environment xrdp running on

VM

### What's your client?

_No response_

### Area(s) with issue?

Compatiblity aginst clients

### Steps to reproduce

A privileged access management (PAM) product "Safeguard for Privileged Sessions" from OneIdentity (brand of [Quest](https://en.wikipedia.org/wiki/Quest_Software)) fails to connect to xrdp. A session is opened and immediately closed before any interaction is possible.

OneIdentity/Quest refuses to support xrdp.

Seemingly relevant log entries from the PAM product:

```
rdp.error(4): (svc/b6mu1FJJRQBizkcnpP1bbz/RDP_Internal:21/rdp): Static virtual channel count mismatch; client_to_server='4', server_to_client='3'
rdp.error(4): (svc/b6mu1FJJRQBizkcnpP1bbz/RDP_Internal:21/rdp): Function failed; function='rdp_mangle_mcs_init_rsp'
rdp.error(4): (svc/b6mu1FJJRQBizkcnpP1bbz/RDP_Internal:21/rdp): Function failed; function='rdp_mangle_pdu_dt'
rdp.error(4): (svc/b6mu1FJJRQBizkcnpP1bbz/RDP_Internal:21/rdp): Function failed; function='rdp_mangle_iso3'
```

Not sure whether this is some protocol version mismatch, different implementations / implemented protocol features or something else.

Please note that in the issue form "Area(s) with issue?" has a small typo - "aginst".

### ✔️ Expected Behavior

Session working.

### ❌ Actual Behavior

Session fails immediately.

### Anything else?

_No response_

With that final version xrdp logs this upon a failed connection via SPS:

```
[2025-04-17T13:49:07.261+0000] [INFO ] Socket 14: connection accepted from [<censored>]:42690
[2025-04-17T13:49:07.265+0000] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[2025-04-17T13:49:07.266+0000] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[2025-04-17T13:49:07.267+0000] [INFO ] Security protocol: configured [SSL|RDP], requested [SSL|HYBRID|HYBRID_EX|RDP], selected [SSL]
[2025-04-17T13:49:07.449+0000] [INFO ] Connected client computer name: <censored>
[2025-04-17T13:49:07.450+0000] [INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x07], keyboard_subtype:[0x00], keylayout:[0x00010426]
[2025-04-17T13:49:07.452+0000] [INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[2025-04-17T13:49:07.453+0000] [WARN ] Physical desktop dimensions (0x0) are invalid
[2025-04-17T13:49:07.460+0000] [ERROR] SSL_read: I/O error
[2025-04-17T13:49:07.462+0000] [ERROR] libxrdp_force_read: header read error
[2025-04-17T13:49:07.463+0000] [ERROR] Processing [ITU-T T.125] ErectDomainRequest failed
[2025-04-17T13:49:07.464+0000] [ERROR] [MCS Connection Sequence] receive erect domain request failed
[2025-04-17T13:49:07.465+0000] [ERROR] xrdp_sec_incoming: xrdp_mcs_incoming failed
[2025-04-17T13:49:07.467+0000] [ERROR] xrdp_rdp_incoming: xrdp_sec_incoming failed
[2025-04-17T13:49:07.470+0000] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[2025-04-17T13:49:07.471+0000] [ERROR] xrdp_iso_send: trans_write_copy_s failed
[2025-04-17T13:49:07.472+0000] [ERROR] Sending [ITU T.125] DisconnectProviderUltimatum failed
```

Edit: But this log seems to be the same with the RPM version of xrdp.

Absolutely nothing to apologise for, that script already was many times more than what I expected - which was "recompile like this, figure out how" :)

Made the config change, but not seeing any lines like that. Checked that the effective xrdp binary includes `--enable-devel-all` in the output of `-v`, the relevant config file section was:

```
[LoggingPerLogger]
xrdp_sec_process_mcs_data_channels()=TRACE
```

There's this comment, but I assume "logging" is a subset of "all".

```
; Note: per logger configuration is only used if xrdp is built with
; --enable-devel-logging
```

Set `LogLevel=TRACE` in the `Logging` section, got this upon a failed connection - which doesn't seem to be much more than before.

```
[2025-04-18T21:27:53.274+0000] [INFO ] Security protocol: configured [SSL|RDP], requested [SSL|HYBRID|HYBRID_EX|RDP], selected [SSL]
[2025-04-18T21:27:53.390+0000] [TRACE] TLS connection accepted
[2025-04-18T21:27:53.391+0000] [DEBUG] Using TLS security, and setting RDP security crypto to LEVEL_NONE and METHOD_NONE
[2025-04-18T21:27:53.392+0000] [DEBUG] [MCS Connection Sequence] receive connection request
[2025-04-18T21:27:53.509+0000] [INFO ] Connected client computer name: <censored>
[2025-04-18T21:27:53.511+0000] [INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x07], keyboard_subtype:[0x00], keylayout:[0x00010426]
[2025-04-18T21:27:53.512+0000] [DEBUG] keyboard_cfg_file /etc/xrdp/xrdp_keyboard.ini
[2025-04-18T21:27:53.513+0000] [DEBUG] xrdp_load_keyboard_layout: item rdp_layout_us value 0x00000409
[2025-04-18T21:27:53.514+0000] [DEBUG] xrdp_load_keyboard_layout: skipping configuration item - rdp_layout_us, continuing to next section
[2025-04-18T21:27:53.515+0000] [DEBUG] xrdp_load_keyboard_layout: item rdp_layout_us value us
[2025-04-18T21:27:53.516+0000] [DEBUG] xrdp_load_keyboard_layout: skipping configuration item - rdp_layout_us, continuing to next section
[2025-04-18T21:27:53.517+0000] [DEBUG] xrdp_load_keyboard_layout: item keyboard_type value 4
[2025-04-18T21:27:53.518+0000] [DEBUG] xrdp_load_keyboard_layout: item keyboard_subtype value 3
[2025-04-18T21:27:53.519+0000] [DEBUG] xrdp_load_keyboard_layout: item rdp_layouts value default_rdp_layouts
[2025-04-18T21:27:53.520+0000] [DEBUG] xrdp_load_keyboard_layout: item layouts_map value rdp_layouts_map_mac
[2025-04-18T21:27:53.521+0000] [DEBUG] xrdp_load_keyboard_layout: item keyboard_type value 7
[2025-04-18T21:27:53.522+0000] [DEBUG] xrdp_load_keyboard_layout: item keyboard_subtype value 2
[2025-04-18T21:27:53.523+0000] [DEBUG] xrdp_load_keyboard_layout: item rdp_layout_us value us
[2025-04-18T21:27:53.524+0000] [DEBUG] xrdp_load_keyboard_layout: skipping configuration item - rdp_layout_us, continuing to next section
[2025-04-18T21:27:53.525+0000] [INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[2025-04-18T21:27:53.526+0000] [WARN ] Physical desktop dimensions (0x0) are invalid
[2025-04-18T21:27:53.527+0000] [DEBUG] Client supports 40 bit encryption
[2025-04-18T21:27:53.528+0000] [DEBUG] Client supports 128 bit encryption
[2025-04-18T21:27:53.529+0000] [DEBUG] Client supports 56 bit encryption
[2025-04-18T21:27:53.530+0000] [DEBUG] Client supports fips encryption
[2025-04-18T21:27:53.531+0000] [DEBUG] The connection is using TLS, skipping RDP crypto negotiation
[2025-04-18T21:27:53.532+0000] [DEBUG] Adding channel: name rdpdr, channel id 1004, flags 0x80800000
[2025-04-18T21:27:53.533+0000] [DEBUG] Adding channel: name cliprdr, channel id 1005, flags 0xc0a00000
[2025-04-18T21:27:53.534+0000] [DEBUG] Adding channel: name drdynvc, channel id 1006, flags 0xc0800000
[2025-04-18T21:27:53.535+0000] [DEBUG] [MCS Connection Sequence] construct connection response
[2025-04-18T21:27:53.536+0000] [DEBUG] using no security
[2025-04-18T21:27:53.537+0000] [DEBUG] [MCS Connection Sequence] send connection response
[2025-04-18T21:27:53.539+0000] [DEBUG] [MCS Connection Sequence] receive erect domain request
[2025-04-18T21:27:53.540+0000] [ERROR] SSL_read: I/O error
[2025-04-18T21:27:53.541+0000] [ERROR] libxrdp_force_read: header read error
[2025-04-18T21:27:53.542+0000] [ERROR] Processing [ITU-T T.125] ErectDomainRequest failed
[2025-04-18T21:27:53.544+0000] [ERROR] [MCS Connection Sequence] receive erect domain request failed
[2025-04-18T21:27:53.545+0000] [ERROR] xrdp_sec_incoming: xrdp_mcs_incoming failed
[2025-04-18T21:27:53.546+0000] [ERROR] xrdp_rdp_incoming: xrdp_sec_incoming failed
[2025-04-18T21:27:53.547+0000] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[2025-04-18T21:27:53.548+0000] [ERROR] xrdp_iso_send: trans_write_copy_s failed
[2025-04-18T21:27:53.549+0000] [DEBUG] Closed socket 14 ([::ffff:172.31.176.40]:3389)
[2025-04-18T21:27:53.550+0000] [ERROR] Sending [ITU T.125] DisconnectProviderUltimatum failed
[2025-04-18T21:27:53.551+0000] [TRACE] [xrdp_process_run(xrdp_listen.c:154)] process done
```

Thanks, that now gives this:

```
[2025-04-22T14:44:59.670+0000] [INFO ] [g_sck_accept(os_calls.c:1253)] Socket 14: connection accepted from [<censored>]:36410
[2025-04-22T14:44:59.672+0000] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:227)] Using default X.509 certificate: /etc/xrdp/cert.pem
[2025-04-22T14:44:59.674+0000] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:258)] Using default X.509 key file: /etc/xrdp/key.pem
[2025-04-22T14:44:59.675+0000] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:196)] Security protocol: configured [SSL|RDP], requested [SSL|HYBRID|HYBRID_EX|RDP], selected [SSL]
[2025-04-22T14:44:59.888+0000] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1771)] Connected client computer name: <censored>
[2025-04-22T14:44:59.890+0000] [INFO ] [xrdp_load_keyboard_layout(xrdp_sec.c:181)] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x07], keyboard_subtype:[0x00], keylayout:[0x00010426]
[2025-04-22T14:44:59.892+0000] [INFO ] [xrdp_load_keyboard_layout(xrdp_sec.c:388)] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[2025-04-22T14:44:59.893+0000] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1914)] client DOES NOT support gfx
[2025-04-22T14:44:59.895+0000] [WARN ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1983)] Physical desktop dimensions (0x0) are invalid
[2025-04-22T14:44:59.896+0000] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2147)] Received [MS-RDPBCGR] TS_UD_CS_NET channelCount 4
[2025-04-22T14:44:59.897+0000] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2176)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 0, name rdpdr, options 0x80800000
[2025-04-22T14:44:59.898+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2182)] Adding channel: name rdpdr, channel id 1004, flags 0x80800000
[2025-04-22T14:44:59.900+0000] [WARN ] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2188)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 1, skipped because of malformed channel name.
[2025-04-22T14:44:59.901+0000] [WARN ] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2191)] [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF name Hex Dump:
0000   5f 5f 5f 5f 5f 5f 5f 5f                           ________        
[2025-04-22T14:44:59.902+0000] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2176)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 2, name cliprdr, options 0xc0a00000
[2025-04-22T14:44:59.903+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2182)] Adding channel: name cliprdr, channel id 1005, flags 0xc0a00000
[2025-04-22T14:44:59.904+0000] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2176)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 3, name drdynvc, options 0xc0800000
[2025-04-22T14:44:59.907+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2182)] Adding channel: name drdynvc, channel id 1006, flags 0xc0800000
[2025-04-22T14:44:59.908+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2200)] MCS user is 6, channel id is 1007
[2025-04-22T14:44:59.910+0000] [ERROR] [ssl_tls_log_error(ssl_calls.c:992)] SSL_read: I/O error
[2025-04-22T14:44:59.912+0000] [ERROR] [libxrdp_force_read(libxrdp.c:134)] libxrdp_force_read: header read error
[2025-04-22T14:44:59.913+0000] [ERROR] [xrdp_mcs_recv_edrq(xrdp_mcs.c:527)] Processing [ITU-T T.125] ErectDomainRequest failed
[2025-04-22T14:44:59.914+0000] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1236)] [MCS Connection Sequence] receive erect domain request failed
[2025-04-22T14:44:59.915+0000] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2604)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[2025-04-22T14:44:59.917+0000] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:941)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[2025-04-22T14:44:59.918+0000] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[2025-04-22T14:44:59.919+0000] [ERROR] [xrdp_iso_send(xrdp_iso.c:639)] xrdp_iso_send: trans_write_copy_s failed
[2025-04-22T14:44:59.920+0000] [ERROR] [xrdp_mcs_disconnect(xrdp_mcs.c:1424)] Sending [ITU T.125] DisconnectProviderUltimatum failed
```

Thank you so much, log with the latest version and after `gmake clean`:

```
[2025-04-26T16:35:20.759+0000] [INFO ] [g_sck_accept(os_calls.c:1253)] Socket 14: connection accepted from [<censored>]:60766
[2025-04-26T16:35:20.761+0000] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:227)] Using default X.509 certificate: /etc/xrdp/cert.pem
[2025-04-26T16:35:20.762+0000] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:258)] Using default X.509 key file: /etc/xrdp/key.pem
[2025-04-26T16:35:20.763+0000] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:196)] Security protocol: configured [SSL|RDP], requested [SSL|HYBRID|HYBRID_EX|RDP], selected [SSL]
[2025-04-26T16:35:21.109+0000] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1771)] Connected client computer name: <censored>
[2025-04-26T16:35:21.122+0000] [INFO ] [xrdp_load_keyboard_layout(xrdp_sec.c:181)] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x07], keyboard_subtype:[0x00], keylayout:[0x00010426]
[2025-04-26T16:35:21.136+0000] [INFO ] [xrdp_load_keyboard_layout(xrdp_sec.c:388)] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[2025-04-26T16:35:21.146+0000] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1914)] client DOES NOT support gfx
[2025-04-26T16:35:21.155+0000] [WARN ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1983)] Physical desktop dimensions (0x0) are invalid
[2025-04-26T16:35:21.166+0000] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2147)] Received [MS-RDPBCGR] TS_UD_CS_NET channelCount 4
[2025-04-26T16:35:21.176+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2181)] Adding channel: name rdpdr, channel id 1004, flags 0x80800000
[2025-04-26T16:35:21.185+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2181)] Adding channel: name _______, channel id 1005, flags 0xc0000000
[2025-04-26T16:35:21.195+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2181)] Adding channel: name cliprdr, channel id 1006, flags 0xc0a00000
[2025-04-26T16:35:21.205+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2181)] Adding channel: name drdynvc, channel id 1007, flags 0xc0800000
[2025-04-26T16:35:21.213+0000] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2188)] MCS user is 7, channel id is 1008
[2025-04-26T16:35:21.877+0000] [INFO ] [xrdp_rdp_incoming(xrdp_rdp.c:955)] TLS connection established from [<censored>]:60766 TLSv1.3 with cipher TLS_AES_256_GCM_SHA384
[2025-04-26T16:35:21.256+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:803)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_GENERAL
[2025-04-26T16:35:21.257+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:808)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_BITMAP
[2025-04-26T16:35:21.259+0000] [INFO ] [xrdp_caps_process_bitmap(xrdp_caps.c:151)] Client supports single-screen resizes by xrdp
[2025-04-26T16:35:21.260+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:813)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_ORDER
[2025-04-26T16:35:21.260+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:818)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_BITMAPCACHE
[2025-04-26T16:35:21.261+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:832)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_ACTIVATION - Ignored
[2025-04-26T16:35:21.262+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:823)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_CONTROL - Ignored
[2025-04-26T16:35:21.263+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:836)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_POINTER
[2025-04-26T16:35:21.264+0000] [INFO ] [xrdp_caps_process_pointer(xrdp_caps.c:391)] xrdp_caps_process_pointer: client supports new(color) cursor
[2025-04-26T16:35:21.265+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:841)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_SHARE - Ignored
[2025-04-26T16:35:21.266+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:853)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_INPUT
[2025-04-26T16:35:21.267+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:849)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_SOUND - Ignored
[2025-04-26T16:35:21.268+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:858)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_FONT - Ignored
[2025-04-26T16:35:21.269+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:867)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_GLYPHCACHE
[2025-04-26T16:35:21.270+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:862)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_BRUSH
[2025-04-26T16:35:21.271+0000] [INFO ] [xrdp_caps_process_confirm_active(xrdp_caps.c:882)] Received [MS-RDPBCGR] TS_CONFIRM_ACTIVE_PDU - TS_CAPS_SET capabilitySetType = CAPSTYPE_VIRTUALCHANNEL - Ignored
[2025-04-26T16:35:21.272+0000] [WARN ] [xrdp_rdp_process_data_control(xrdp_rdp.c:1173)] Received [MS-RDPBCGR] TS_CONTROL_PDU action 4 is unknown (skipped)
[2025-04-26T16:35:21.273+0000] [INFO ] [xrdp_rdp_process_data_font(xrdp_rdp.c:1310)] yeah, up_and_running
[2025-04-26T16:35:21.274+0000] [INFO ] [xrdp_bitmap_create(xrdp_bitmap_common.c:102)] xrdp_bitmap_create: noorders
[2025-04-26T16:35:21.279+0000] [INFO ] [xrdp_mm_create(xrdp_mm.c:68)] xrdp_mm_create: bpp 16 mcs_connection_type 7 jpeg_codec_id 0 v3_codec_id 0 rfx_codec_id 0 h264_codec_id 0
[2025-04-26T16:35:21.280+0000] [WARN ] [get_keymaps(lang.c:246)] Cannot find keymap file /etc/xrdp/km-00010426.ini
[2025-04-26T16:35:21.281+0000] [WARN ] [get_keymaps(lang.c:254)] Cannot find keymap file /etc/xrdp/km-00000426.ini
[2025-04-26T16:35:21.282+0000] [INFO ] [km_load_file(lang.c:290)] Loading keymap file /etc/xrdp/km-00000409.ini
[2025-04-26T16:35:21.284+0000] [WARN ] [get_keymaps(lang.c:271)] local keymap file for 0x00010426 found and doesn't match built in keymap, using local keymap file
[2025-04-26T16:35:21.285+0000] [WARN ] [xrdp_wm_load_channel_config(xrdp_wm.c:74)] Static channel '_______' from the client is not named in the [Channels] section
[2025-04-26T16:35:21.286+0000] [INFO ] [xrdp_channel_drdynvc_start(xrdp_channel.c:758)] xrdp_channel_drdynvc_start: drdynvc_channel_id -1
[2025-04-26T16:35:21.288+0000] [INFO ] [tconfig_load_gfx(xrdp_tconfig.c:524)] TConfig: Loading GFX config file /etc/xrdp/gfx.toml
[2025-04-26T16:35:21.289+0000] [WARN ] [xrdp_login_wnd_get_monitor_dpi(xrdp_login_wnd.c:706)] No information is available to determine login screen DPI
[2025-04-26T16:35:21.290+0000] [WARN ] [xrdp_font_create(xrdp_font.c:164)] No DPI value is available to find login font
[2025-04-26T16:35:21.291+0000] [WARN ] [xrdp_font_create(xrdp_font.c:166)] Using the default_dpi of 96
[2025-04-26T16:35:21.298+0000] [WARN ] Ignoring obsolete SCP port value '3350'
[2025-04-26T16:35:21.299+0000] [INFO ] [xrdp_wm_log_msg(xrdp_wm.c:2345)] xrdp_wm_log_msg: sesman connect ok
[2025-04-26T16:35:21.300+0000] [INFO ] [xrdp_wm_log_msg(xrdp_wm.c:2345)] xrdp_wm_log_msg: Performing login request for <censored>
[2025-04-26T16:35:21.364+0000] [INFO ] [xrdp_wm_log_msg(xrdp_wm.c:2345)] xrdp_wm_log_msg: login was successful - creating session
[2025-04-26T16:35:21.426+0000] [ERROR] [dynamic_monitor_open_response(xrdp_mm.c:1138)] dynamic_monitor_open_response: error
[2025-04-26T16:35:21.428+0000] [ERROR] [xrdp_rdp_recv(xrdp_rdp.c:506)] xrdp_rdp_recv: xrdp_channel_process failed
[2025-04-26T16:35:22.546+0000] [INFO ] [xrdp_wm_log_msg(xrdp_wm.c:2345)] xrdp_wm_log_msg: session is available on display 10 for user <censored>
[2025-04-26T16:35:22.547+0000] [INFO ] [xrdp_wm_log_msg(xrdp_wm.c:2345)] xrdp_wm_log_msg: Connecting to session
[2025-04-26T16:35:22.550+0000] [INFO ] [xrdp_mm_setup_mod1(xrdp_mm.c:356)] loaded module 'libvnc.so' ok, interface size 4832, version 4
[2025-04-26T16:35:22.553+0000] [INFO ] VNC: Clipboard (if available) is provided by chansrv facility
[2025-04-26T16:35:22.554+0000] [INFO ] [xrdp_wm_log_msg(xrdp_wm.c:2345)] xrdp_wm_log_msg: Connecting to chansrv
[2025-04-26T16:35:22.593+0000] [INFO ] Layout from ClientLayout (geom=1568x1031 #screens=1) : 0:(1568x1031+0+0)
[2025-04-26T16:35:22.595+0000] [INFO ] Layout from OldServerLayout (geom=1568x1031 #screens=1) : 1804289383:(1568x1031+0+0)
[2025-04-26T16:35:22.596+0000] [INFO ] [xrdp_cache_add_bitmap(xrdp_cache.c:376)] xrdp_cache_add_bitmap: reset detected cache_id 2
[2025-04-26T16:35:22.620+0000] [INFO ] [xrdp_cache_add_bitmap(xrdp_cache.c:376)] xrdp_cache_add_bitmap: reset detected cache_id 1
[2025-04-26T16:35:22.622+0000] [INFO ] [xrdp_cache_add_bitmap(xrdp_cache.c:376)] xrdp_cache_add_bitmap: reset detected cache_id 0
```

--- a/libxrdp/libxrdp.c
+++ b/libxrdp/libxrdp.c
@@ -1297,7 +1297,7 @@ libxrdp_query_channel(struct xrdp_session *session, int channel_id,
 
     if (channel_name != 0)
     {
-        g_strncpy(channel_name, channel_item->name, 8);
+        strlcpy(channel_name, channel_item->name, CHANNEL_NAME_LEN + 1);
         LOG(LOG_LEVEL_DEBUG, "libxrdp_query_channel - Channel %d name %s",
             channel_id, channel_name);
     }

--- a/libxrdp/libxrdp.h
+++ b/libxrdp/libxrdp.h
@@ -50,7 +50,7 @@ struct xrdp_iso
 /* used in mcs */
 struct mcs_channel_item
 {
-    char name[16];
+    char name[CHANNEL_NAME_LEN + 1];
     int flags;
     int chanid;
     int disabled;

--- a/libxrdp/xrdp_sec.c
+++ b/libxrdp/xrdp_sec.c
@@ -1924,31 +1924,19 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
             g_free(channel_item);
             return 1;
         }
-        in_uint8a(s, channel_item->name, 8);
+
+        in_uint8a(s, channel_item->name, CHANNEL_NAME_LEN + 1);
+        // The channel name *should* be null-terminated. Add a back-stop
+        // terminator in case it isn't.
+        channel_item->name[CHANNEL_NAME_LEN] = '\0';
         in_uint32_le(s, channel_item->flags);
 
-        if (g_strlen(channel_item->name) > 0 && g_strlen(channel_item->name) < 8)
-        {
-            LOG_DEVEL(LOG_LEVEL_TRACE, "Received [MS-RDPBCGR] "
-                      "TS_UD_CS_NET.CHANNEL_DEF %d, name %s, options 0x%8.8x",
-                      index, channel_item->name, channel_item->flags);
-            channel_item->chanid = next_mcs_channel_id++;
-            list_add_item(self->mcs_layer->channel_list,
-                          (intptr_t) channel_item);
-            LOG(LOG_LEVEL_DEBUG,
-                "Adding channel: name %s, channel id %d, flags 0x%8.8x",
-                channel_item->name, channel_item->chanid, channel_item->flags);
-        }
-        else
-        {
-            LOG_DEVEL(LOG_LEVEL_WARNING, "Received [MS-RDPBCGR] "
-                      "TS_UD_CS_NET.CHANNEL_DEF %d, skipped because of "
-                      "malformed channel name.", index);
-            LOG_DEVEL_HEXDUMP(LOG_LEVEL_WARNING,
-                              "[MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF name",
-                              channel_item->name, 8);
-            g_free(channel_item);
-        }
+        channel_item->chanid = next_mcs_channel_id++;
+        list_add_item(self->mcs_layer->channel_list,
+                      (intptr_t) channel_item);
+        LOG(LOG_LEVEL_DEBUG,
+            "Adding channel: name %s, channel id %d, flags 0x%8.8x",
+            channel_item->name, channel_item->chanid, channel_item->flags);
     }
 
     /* Set the user channel as well */

--- a/neutrinordp/xrdp-neutrinordp.c
+++ b/neutrinordp/xrdp-neutrinordp.c
@@ -1800,7 +1800,7 @@ lfreerdp_pre_connect(freerdp *instance)
     int num_chans;
     int target_chan;
     int ch_flags;
-    char ch_name[256];
+    char ch_name[CHANNEL_NAME_LEN + 1];
     const char *ch_names[MAX_FREERDP_CHANNELS];
     char *dst_ch_name;
 

--- a/sesman/chansrv/chansrv.c
+++ b/sesman/chansrv/chansrv.c
@@ -371,7 +371,7 @@ process_message_channel_setup(struct stream *s)
     {
         ci = &(g_chan_items[g_num_chan_items]);
         g_memset(ci->name, 0, sizeof(ci->name));
-        in_uint8a(s, ci->name, 8);
+        in_uint8a(s, ci->name, CHANNEL_NAME_LEN + 1);
         in_uint16_le(s, ci->id);
         in_uint16_le(s, ci->flags);
         LOG_DEVEL(LOG_LEVEL_DEBUG, "process_message_channel_setup: chan name '%s' "

--- a/xrdp/xrdp_mm.c
+++ b/xrdp/xrdp_mm.c
@@ -634,9 +634,7 @@ xrdp_mm_trans_send_channel_setup(struct xrdp_mm *self, struct trans *trans)
     int chan_flags;
     int size;
     struct stream *s;
-    char chan_name[256];
-
-    g_memset(chan_name, 0, sizeof(char) * 256);
+    char chan_name[CHANNEL_NAME_LEN + 1];
 
     s = trans_get_out_s(trans, 8192);
 
@@ -656,7 +654,7 @@ xrdp_mm_trans_send_channel_setup(struct xrdp_mm *self, struct trans *trans)
         if (libxrdp_query_channel(self->wm->session, chan_id, chan_name,
                                   &chan_flags) == 0)
         {
-            out_uint8a(s, chan_name, 8);
+            out_uint8a(s, chan_name, CHANNEL_NAME_LEN + 1);
             out_uint16_le(s, chan_id);
             out_uint16_le(s, chan_flags);
             ++output_chan_count;

--- a/xrdp/xrdp_wm.c
+++ b/xrdp/xrdp_wm.c
@@ -48,11 +48,12 @@ xrdp_wm_load_channel_config(struct xrdp_wm *self)
 
         for (chan_id = 0 ; chan_id < chan_count ; ++chan_id)
         {
-            char chan_name[16];
+            char chan_name[CHANNEL_NAME_LEN + 1];
             if (libxrdp_query_channel(self->session, chan_id, chan_name,
                                       NULL) == 0)
             {
                 int disabled = 1; /* Channels disabled if not found */
+                int found = 0;
                 int index;
 
                 for (index = 0; index < names->count; index++)
@@ -61,9 +62,17 @@ xrdp_wm_load_channel_config(struct xrdp_wm *self)
                     const char *r = (const char *)list_get_item(values, index);
                     if (g_strcasecmp(q, chan_name) == 0)
                     {
+                        found = 1;
                         disabled = !g_text2bool(r);
                         break;
                     }
+                }
+                if (!found)
+                {
+                    LOG(LOG_LEVEL_WARNING,
+                        "Static channel '%s' from the client"
+                        " is not named in the [Channels] section",
+                        chan_name);
                 }
                 disabled_str = (disabled) ? "disabled" : "enabled";
                 LOG(LOG_LEVEL_DEBUG, "xrdp_wm_load_channel_config: "

